radix-engine/src/model/method_authorization.rs

```diff
@@ -2,105 +2,164 @@ use crate::errors::RuntimeError;
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash, TypeId, Encode, Decode)]
+pub enum HardProofRuleResourceList {
+    List(Vec<HardResourceOrNonFungible>),
+    SoftResourceListNotFound,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash, TypeId, Encode, Decode)]
+pub enum HardProofRule {
+    This(HardResourceOrNonFungible),
+    SomeOfResource(Decimal, HardResourceOrNonFungible),
+    AllOf(HardProofRuleResourceList),
+    AnyOf(HardProofRuleResourceList),
+    CountOf(u8, HardProofRuleResourceList),
+}
 
-                Err(NotAuthorized)
+impl HardProofRule {
+    pub fn check(&self, proofs_vector: &[&[Proof]]) -> Result<(), RuntimeError> {
+        match self {
+            HardProofRule::This(resource) => {
+                if resource.check(proofs_vector) {
+                    Ok(())
+                } else {
+                    Err(NotAuthorized)
+                }
             }
-            ProofRule::AllOf(rules) => {
-                for rule in rules {
-                    if Self::check_proof_rule(rule, proofs_vector).is_err() {
-                        return Err(NotAuthorized);
-                    }
+            HardProofRule::SomeOfResource(amount, resource) => {
+                if resource.check_has_amount(*amount, proofs_vector) {
+                    Ok(())
+                } else {
+                    Err(NotAuthorized)
                 }
-
-                Ok(())
             }
-            ProofRule::OneOf(rules) => {
-                for rule in rules {
-                    if Self::check_proof_rule(rule, proofs_vector).is_ok() {
-                        return Ok(());
+            HardProofRule::AllOf(resource_list) => match resource_list {
+                HardProofRuleResourceList::SoftResourceListNotFound => Err(NotAuthorized),
+                HardProofRuleResourceList::List(resources) => {
+                    for resource in resources {
+                        if !resource.check(proofs_vector) {
+                            return Err(NotAuthorized);
+                        }
                     }
-                }
 
-                Err(NotAuthorized)
-            }
-            ProofRule::CountOf { count, rules } => {
-                let mut left = count.clone();
-                for rule in rules {
-                    if Self::check_proof_rule(rule, proofs_vector).is_ok() {
-                        left -= 1;
-                        if left == 0 {
+                    Ok(())
+                }
+            },
+            HardProofRule::AnyOf(resource_list) => match resource_list {
+                HardProofRuleResourceList::SoftResourceListNotFound => Err(NotAuthorized),
+                HardProofRuleResourceList::List(resources) => {
+                    for resource in resources {
+                        if resource.check(proofs_vector) {
                             return Ok(());
                         }
                     }
+
+                    Err(NotAuthorized)
```

Comment:
We might have discussed during team meeting, but it's a bit hard for me to accept `AllOf([]) == Ok` and `AnyOf([]) == NotAuthorized` without reading the implementation.

Do you think it's worth to add a "at least one" constraint on the list, for both static and dynamic ones?

Reply:
Sure, I can replace `AnyOf` with `AtleastOneOf`. Is that what you mean?

Reply:
Sorry, I was trying to avoid the empty set condition by issuing a `SoftResourceNotFound` if the resource list is empty. 

Now I'm thinking `AllOf` as logical `for all` and `AnyOf` as logical `there exists`. 
- The formula  `\forall {x}{\in}\emptyset P(x)}` is always true, regardless of the predicate P(x)
- The formula `\exists {x}{\in}\emptyset P(x)}` is always false, regardless of the predicate P(x)

The current implementation starts to make sense. Let's keep it as is.

Reply:
yup exactly